pkg/csi/manila/nodeserver.go

```diff
@@ -0,0 +1,329 @@
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Internal, "failed to retrieve export locations for volume %s (share ID %s): %v",
+			volID, share.ID, err)
+	}
+
+	chosenExportLocation, err := chooseExportLocation(availableExportLocations)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Internal, "failed to choose an export locations for volume %s (share ID %s): %v",
+			volID, share.ID, err)
+	}
+
+	accessRights, err := shares.ListAccessRights(manilaClient, share.ID).Extract()
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Internal, "failed to list access rights for volume %s (share ID %s): %v",
+			volID, share.ID, err)
+	}
+
+	// Get the access right for this share
+
+	var accessRight *shares.AccessRight
+	for i := range accessRights {
+		if accessRights[i].ID == shareOpts.ShareAccessID {
+			accessRight = &accessRights[i]
+			break
+		}
+	}
+
+	if accessRight == nil {
+		return nil, nil, status.Errorf(codes.InvalidArgument, "cannot find access right %s for volume %s (share ID %s)",
+			shareOpts.ShareAccessID, volID, share.ID)
+	}
+
+	// Build volume context and secrets for fwd plugin
+
+	sa := getShareAdapter(ns.d.shareProto)
+
+	volumeContext, err = sa.BuildVolumeContext(&shareadapters.VolumeContextArgs{Location: &chosenExportLocation, Options: shareOpts})
+	if err != nil {
+		return nil, nil, status.Errorf(codes.InvalidArgument, "failed to build volume context for volume %s: %v", volID, err)
+	}
+
+	secrets, err = sa.BuildSecrets(&shareadapters.SecretsArgs{AccessRight: accessRight})
+	if err != nil {
+		return nil, nil, status.Errorf(codes.InvalidArgument, "failed to build volume secrets for volume %s: %v", volID, err)
+	}
+
+	return
+}
+
+func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	if err := validateNodePublishVolumeRequest(req); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	// Configuration
+
+	shareOpts, err := options.NewNodeVolumeContext(req.GetVolumeContext())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid volume parameters: %v", err)
+	}
+
+	osOpts, err := options.NewOpenstackOptions(req.GetSecrets())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid volume secrets: %v", err)
+	}
+
+	volID := newVolumeID(req.GetVolumeId())
+
+	var volumeCtx, secrets map[string]string
+
+	if ns.supportsNodeStage {
+		// STAGE_UNSTAGE_VOLUME capability is enabled, NodeStageVolume should've already built the staging data
+
+		ns.nodeStageCacheMtx.RLock()
+		cacheEntry, ok := ns.nodeStageCache[volID]
+		ns.nodeStageCacheMtx.RUnlock()
+
+		if ok {
+			volumeCtx, secrets = cacheEntry.volumeContext, cacheEntry.secrets
+		} else {
+			klog.Errorf("STAGE_UNSTAGE_VOLUME capability is enabled, but node stage cache doesn't contain an entry for %s - this is most likely a bug! Rebuild staging data anyway...")
```

Comment:
And once more: missing argument for the string format.

Reply:
Thanks for noticing those!

Reply:
Fixed in https://github.com/kubernetes/cloud-provider-openstack/pull/536/commits/91be405578e36c59d7f69f1c6afdbb364348989f